sql/kvstore/kvstore.go

```diff
@@ -0,0 +1,62 @@
+package kvstore
+
+import (
+	"fmt"
+
+	"github.com/spacemeshos/go-scale"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+func addKeyValue(db sql.Executor, key string, value scale.Encodable) error {
+	bytes, err := codec.Encode(value)
+	if err != nil {
+		return fmt.Errorf("failed encoding: %w", err)
+	}
+
+	if _, err := db.Exec(`
+		insert into kvstore (id, value) values (?1, ?2)
+		on conflict (id) do
+		update set value = ?2;`,
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, []byte(key))
+			stmt.BindBytes(2, bytes)
+		}, nil); err != nil {
+		return fmt.Errorf("failed to insert value: %w", err)
+	}
+
+	return nil
+}
+
+func getKeyValue(db sql.Executor, key string, value scale.Decodable) error {
+	var val []byte
+	if rows, err := db.Exec("select value from kvstore where id = ?1;", func(stmt *sql.Statement) {
+		stmt.BindBytes(1, []byte(key))
+	}, func(stmt *sql.Statement) bool {
+		val = make([]byte, stmt.ColumnLen(0))
+		stmt.ColumnBytes(0, val[:])
+		return true
+	}); err != nil {
+		return fmt.Errorf("failed to get value: %w", err)
+	} else if rows == 0 {
+		return fmt.Errorf("failed to get value: %w", sql.ErrNotFound)
+	}
+
+	if len(val) == 0 {
+		return nil
+	}
```

Comment:
Is this check really required?  This case should be handled by `codec.Decode`, shouldn't it?

Reply:
Not quite: If the value is empty for a given key it will return no error. This is the behaviour as it was before the refactoring.

If I remove this check it will instead return an error (a wrapped `io.EOF` because `codec.Decode` tries to read from a buffer that is empty).

I'm not sure what is the correct behaviour here. I'll remove this check and see if anything breaks